File: android/src/main/kotlin/com/example/confidence_flutter_sdk/ConfidenceFlutterSdkPlugin.kt
package com.example.confidence_flutter_sdk

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.spotify.confidence.Confidence
import com.spotify.confidence.ConfidenceFactory
import com.spotify.confidence.ConfidenceValue
import com.spotify.confidence.LoggingLevel
import com.spotify.confidence.FlagResolution
import com.spotify.confidence.client.SdkMetadata
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import java.io.File
import java.util.concurrent.atomic.AtomicBoolean

private const val TAG = "ConfidenceFlutterSdk"

/**
 * Replies to a [Result] exactly once, on the platform (main) thread.
 *
 * Flutter requires channel replies on the main thread, and a second reply
 * throws. The async cases in this plugin complete on [Dispatchers.IO], so
 * neither guarantee holds at the call site — both are enforced here.
 */
private class MainThreadResult(private val delegate: Result) {
  private val replied = AtomicBoolean(false)

  fun success(value: Any?) = replyOnce { delegate.success(value) }

  fun error(code: String, message: String?) = replyOnce { delegate.error(code, message, null) }

  private fun replyOnce(reply: () -> Unit) {
    if (!replied.compareAndSet(false, true)) return
    if (Looper.myLooper() == Looper.getMainLooper()) {
      reply()
    } else {
      Handler(Looper.getMainLooper()).post(reply)
    }
  }
}

/** ConfidenceFlutterSdkPlugin */
class ConfidenceFlutterSdkPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel
  private lateinit var confidence: Confidence
  private val coroutineScope = CoroutineScope(Dispatchers.IO)
  private lateinit var context: Context

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "confidence_flutter_sdk")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when(call.method) {
      "flush" -> {
        try {
          confidence.flush()
          result.success(null)
        } catch (e: Exception) {
          result.error("FLUSH_FAILED", "Failed to flush: ${e.message}", null)
        }
      }
      "setup" -> {
        val apiKey = call.argument<String>("apiKey")!!
        val loggingLevel = call.argument<String>("loggingLevel")!!
        val resolveBaseUrl = call.argument<String>("resolveBaseUrl")
        val resolvedLoggingLevel = LoggingLevel.valueOf(loggingLevel)
        confidence = if (resolveBaseUrl.isNullOrBlank()) {
          ConfidenceFactory.create(
            context,
            apiKey,
            loggingLevel = resolvedLoggingLevel
          )
        } else {
          ConfidenceFactory.create(
            context,
            apiKey,
            resolveBaseUrl = resolveBaseUrl,
            loggingLevel = resolvedLoggingLevel
          )
        }
        result.success(null)
      }
      "fetchAndActivate" -> {
        val reply = MainThreadResult(result)
        coroutineScope.launch {
          try {
            confidence.fetchAndActivate()
            reply.success(null)
          } catch (e: Throwable) {
            Log.e(TAG, "fetchAndActivate failed", e)
            reply.error("FETCH_AND_ACTIVATE_FAILED", e.message)
          }
        }
      }
      "activateAndFetchAsync" -> {
        val reply = MainThreadResult(result)
        coroutineScope.launch {
          try {
            confidence.activate()
            confidence.asyncFetch()
            reply.success(null)
          } catch (e: Throwable) {
            Log.e(TAG, "activateAndFetchAsync failed", e)
            reply.error("ACTIVATE_AND_FETCH_ASYNC_FAILED", e.message)
          }
        }
      }
      "isStorageEmpty" -> {
        val isEmpty = confidence.isStorageEmpty()
        result.success(isEmpty)
      }
      "getString" -> {
        val key = call.argument<String>("key")!!
        val defaultValue = call.argument<String>("defaultValue")
        val value = confidence.getValue(key, defaultValue)
        result.success(value)
      }
      "getDouble" -> {
        val key = call.argument<String>("key")!!
        val defaultValue = call.argument<Double>("defaultValue")
        val value = confidence.getValue(key, defaultValue)
        result.success(value)
      }
      "getBool" -> {
        val key = call.argument<String>("key")!!
        val defaultValue = call.argument<Boolean>("defaultValue")
        val value = confidence.getValue(key, defaultValue)
        result.success(value)
      }
      "getInt" -> {
        val key = call.argument<String>("key")!!
        val defaultValue = call.argument<Int>("defaultValue")
        val value = confidence.getValue(key, defaultValue)
        result.success(value)
      }
      "getObject" -> {
        val key = call.argument<String>("key")!!
        val wrappedDefaultValue = call.argument<Map<String, Map<String, Any>>>("defaultValue")!!
        val defaultValue: ConfidenceValue.Struct = ConfidenceValue.Struct(wrappedDefaultValue.mapValues { (_, value) -> value.convert() })
        val value = confidence.getValue(key, defaultValue)
        result.success(Json.encodeToString(NetworkConfidenceValueSerializer, value))
      }
      "readAllFlags" -> {
        val reply = MainThreadResult(result)
        coroutineScope.launch {
          try {
            val flags = readAllFlags()
            val map = flags.flags.associateBy({ it.flag }, { ConfidenceValue.Struct(it.value) })
            reply.success(Json.encodeToString(NetworkConfidenceValueSerializer, ConfidenceValue.Struct(map)))
          } catch (e: Throwable) {
            // A corrupt or partially written cache file makes decoding throw;
            // without a reply the Dart future would never complete.
            Log.e(TAG, "readAllFlags failed", e)
            reply.error("READ_ALL_FLAGS_FAILED", e.message)
          }
        }
      }
      "putContext" -> {
        val key = call.argument<String>("key")!!
        val value = call.argument<Map<String, Any>>("value")!!.convert()
        confidence.putContext(key, value)
        result.success(null)
      }
      "putAllContext" -> {
        val wrappedContext = call.argument<Map<String, Map<String, Any>>>("context")!!
        val context: Map<String, ConfidenceValue> = wrappedContext.mapValues { (_, value) -> value.convert() }
        confidence.putContext(context)
        result.success(null)
      }
      "track" -> {
        val eventName = call.argument<String>("eventName")!!
        val wrappedData = call.argument<Map<String, Map<String, Any>>>("data")!!
        val data: Map<String, ConfidenceValue> = wrappedData.mapValues { (_, value) -> value.convert() }
        try {
          confidence.track(eventName, data)
          result.success(null)
        } catch (e: Exception) {
          result.error("TRACK_FAILED", "Failed to track event '$eventName': ${e.message}", null)
        }
      }
      else -> result.notImplemented()
    }
  }

  private fun readAllFlags(): FlagResolution {
    val flagsFile = File(context.filesDir, "confidence_flags_cache.json")
    if (!flagsFile.exists()) return FlagResolution.EMPTY
    val fileText: String = flagsFile.bufferedReader().use { it.readText() }
    return if (fileText.isEmpty()) {
      FlagResolution.EMPTY
    } else {
      Json.decodeFromString(fileText)
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    context = binding.applicationContext
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    context = binding.activity.applicationContext
  }

  override fun onDetachedFromActivityForConfigChanges() {

  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {

  }

  override fun onDetachedFromActivity() {

  }
}

private fun Map<String, Any>.convert(): ConfidenceValue {
  when(val type = this["type"] as String) {
    "string" -> return ConfidenceValue.String(this["value"] as String)
    "double" -> return ConfidenceValue.Double(this["value"] as Double)
    "bool" -> return ConfidenceValue.Boolean(this["value"] as Boolean)
    "int" -> return ConfidenceValue.Integer(this["value"] as Int)
    // Dart could not map this type (a DateTime, a null, a custom object) and
    // has already sent `value.toString()`. Keep that string rather than
    // throwing: `track` is fire-and-forget from Dart, so an exception here
    // would surface only as a logged handler error while the event is lost.
    "unknown" -> return ConfidenceValue.String(this["value"]?.toString() ?: "")
    "list" -> {
      val list = (this["value"] as List<Map<String, Any>>).map { it.convert() }
      return ConfidenceValue.List(list)
    }
    "map" -> {
      val objectValue = this["value"] as Map<String, Any>
      val map = mutableMapOf<String, ConfidenceValue>()
      for((key, value) in objectValue) {
        map[key] = (value as Map<String, Any>).convert()
      }
      return ConfidenceValue.Struct(map)
    }

    else -> throw IllegalArgumentException("Unknown type $type")
  }
}

File: ios/confidence_flutter_sdk/Sources/confidence_flutter_sdk/ConfidenceFlutterSdkPlugin.swift
import Flutter
import UIKit

/// Replies to a `FlutterResult` exactly once, on the main thread.
///
/// Flutter requires channel replies on the main thread, and a second reply
/// traps. The `Task {}` cases in this plugin resume on an arbitrary executor,
/// so neither guarantee holds at the call site — both are enforced here.
final class MainThreadResult {
    private let delegate: FlutterResult
    private var replied = false
    private let lock = NSLock()

    init(_ delegate: @escaping FlutterResult) {
        self.delegate = delegate
    }

    func reply(_ value: Any?) {
        lock.lock()
        let alreadyReplied = replied
        replied = true
        lock.unlock()
        if alreadyReplied { return }

        if Thread.isMainThread {
            delegate(value)
        } else {
            DispatchQueue.main.async { self.delegate(value) }
        }
    }
}

public class ConfidenceFlutterSdkPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "confidence_flutter_sdk", binaryMessenger: registrar.messenger())
        let instance = ConfidenceFlutterSdkPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    var confidence: Confidence? = nil

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "flush":
            guard let confidence = self.confidence else {
                result("")
                return
            }
            confidence.flush()
            // Confidence.flush() is non-throwing, so there is nothing to catch
            // here — but the reply is still mandatory: without it the Dart
            // future never completes.
            result("")
            break;
        case "readAllFlags":
            guard let flags = try? readAllFlags() else {
                result("{}")
                return
            }
            let map = flags.reduce(into: [String: ConfidenceValue]()) { map, flag in
                map[flag.flag] = flag.value
            }
            let networkMessage = TypeMapper.convert(structure: map)
            let encoder = JSONEncoder()
            let data = try! encoder.encode(networkMessage)
            let str = String(decoding: data, as: UTF8.self)
            result(str)
            break;
        case "setup":
            guard let args = call.arguments as? Dictionary<String, Any> else {
                result("")
                return
            }
            let apiKey = args["apiKey"] as! String
            let logLevel = args["loggingLevel"] as! String
            var builder = Confidence.Builder(clientSecret: apiKey, loggerLevel: loggerLevel(from: logLevel))
            if let resolveBaseUrl = args["resolveBaseUrl"] as? String, !resolveBaseUrl.isEmpty {
                builder = builder.withResolveBaseUrl(resolveBaseUrl: resolveBaseUrl)
            }
            self.confidence = builder.build()
            result("")
            break;
        case "isStorageEmpty":
            guard let confidence = self.confidence else {
                result(true)
                return
            }
            result(confidence.isStorageEmpty())
            break;
        case "fetchAndActivate":
            let fetchReply = MainThreadResult(result)
            Task {
                guard let confidence = self.confidence else {
                    fetchReply.reply("")
                    return
                }
                do {
                    try await confidence.fetchAndActivate()
                    fetchReply.reply("")
                } catch {
                    NSLog("%@", "Confidence SDK: fetchAndActivate failed: \(error)")
                    fetchReply.reply(
                        FlutterError(
                            code: "FETCH_AND_ACTIVATE_FAILED",
                            message: "\(error)",
                            details: nil))
                }
            }
            break;
        case "activateAndFetchAsync":
            let activateReply = MainThreadResult(result)
            Task {
                guard let confidence = self.confidence else {
                    activateReply.reply("")
                    return
                }
                do {
                    try confidence.activate()
                } catch {
                    NSLog("%@", "Confidence SDK: activate failed: \(error)")
                    activateReply.reply(
                        FlutterError(
                            code: "ACTIVATE_AND_FETCH_ASYNC_FAILED",
                            message: "\(error)",
                            details: nil))
                    return
                }
                // Deliberately not awaited: asyncFetch refreshes in the
                // background and its outcome is not part of this reply.
                Task {
                    await confidence.asyncFetch()
                }
                activateReply.reply("")
            }
            break;
        case "putContext":
            guard let args = call.arguments as? Dictionary<String, Any> else {
                result("")
                return
            }
            let key = args["key"] as! String
            let wrappedValue = args["value"] as! Dictionary<String, Any>
            let type = wrappedValue["type"] as! String
            let value = convertValue(type, wrappedValue["value"]!)
            confidence?.putContext(key: key, value: value)
            result("")
            break;
        case "putAllContext":
            guard let args = call.arguments as? Dictionary<String, Dictionary<String, Any>> else {
                result("")
                return
            }
            let context = args["context"] as! Dictionary<String, Dictionary<String, Any>>
            let map: ConfidenceStruct = context.mapValues { wrappedValue in
                let type = wrappedValue["type"] as! String
                return convertValue(type, wrappedValue["value"]!)
            }
            confidence?.putContext(context: map)
            result("")
            break;
        case "track":
            guard let args = call.arguments as? Dictionary<String, Any> else {
                result("")
                return
            }
            let eventName = args["eventName"] as! String
            let data = args["data"] as! Dictionary<String, Dictionary<String, Any>>
            let convertedData = data.convert()
            do {
                try confidence?.track(eventName: eventName, data: convertedData)
                result("")
            } catch {
                NSLog("%@", "Confidence SDK: \(error)")
                result(FlutterError(
                    code: "TRACK_FAILED",
                    message: "Failed to track event '\(eventName)': \(error)",
                    details: nil))
            }
            break;
        case "getBool":
            let arguments = call.arguments as! Dictionary<String, Any>
            let defaultValue = arguments["defaultValue"] as! Bool
            let key = arguments["key"] as! String
            guard let confidence = self.confidence else {
                result(defaultValue)
                return
            }
            let message: Bool = confidence.getValue(key: key, defaultValue: defaultValue)
            result(message)
            break;
        case "getString":
            let arguments = call.arguments as! Dictionary<String, Any>
            let defaultValue = arguments["defaultValue"] as! String
            let key = arguments["key"] as! String
            guard let confidence = self.confidence else {
                result(defaultValue)
                return
            }

            let message: String = confidence.getValue(key: key, defaultValue: defaultValue)
            result(message)
            break;
        case "getDouble":
            let arguments = call.arguments as! Dictionary<String, Any>
            let defaultValue = arguments["defaultValue"] as! Double
            let key = arguments["key"] as! String
            guard let confidence = self.confidence else {
                result(defaultValue)
                return
            }

            let message: Double = confidence.getValue(key: key, defaultValue: defaultValue)
            result(message)
            break;
        case "getInt":
            let arguments = call.arguments as! Dictionary<String, Any>
            let defaultValue = arguments["defaultValue"] as! Int
            let key = arguments["key"] as! String
            guard let confidence = self.confidence else {
                result(defaultValue)
                return
            }

            let message: Int = confidence.getValue(key: key, defaultValue: defaultValue)
            result(message)
            break;
        case "getObject":
            let arguments = call.arguments as! Dictionary<String, Any>
            let defaultValueWrapped = arguments["defaultValue"] as! Dictionary<String, Dictionary<String, Any>>
            let defaultValue = defaultValueWrapped.convert()
            let key = arguments["key"] as! String
            guard let confidence = self.confidence else {
                result([:])
                return
            }

            let message: ConfidenceStruct = confidence.getValue(key: key, defaultValue: defaultValue)
            let networkMessage = TypeMapper.convert(structure: message)
            let encoder = JSONEncoder()
            let data = try! encoder.encode(networkMessage)
            let str = String(decoding: data, as: UTF8.self)
            result(str)
            break;
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    func loggerLevel(from string: String) -> LoggerLevel {
        switch string.uppercased() {
        case "VERBOSE":
            return .TRACE
        case "DEBUG":
            return .DEBUG
        case "WARN":
            return .WARN
        case "ERROR":
            return .ERROR
        default:
            return .WARN
        }
    }
}

func readAllFlags() throws -> [ResolvedValue] {
    let storage = DefaultStorage(filePath: "confidence.flags.resolve")
    let savedFlags = try storage.load(defaultValue: FlagResolution.EMPTY)
    return savedFlags.flags
}

extension Dictionary<String, Dictionary<String, Any>> {
    func convert() -> ConfidenceStruct {
        var map: ConfidenceStruct = [:]
        for (key, wrappedValue) in self {
            let type = wrappedValue["type"] as! String
            map[key] = convertValue(type, wrappedValue["value"]!)
        }
        return map
    }
}

func convertValue(_ type: String, _ value: Any) -> ConfidenceValue {
    switch type {
    case "bool":
        return ConfidenceValue.init(boolean: value as! Bool)
    case "double":
        return ConfidenceValue.init(double: value as! Double)
    case "int":
        return ConfidenceValue.init(integer: value as! Int)
    case "map":
        let dataMap = value as! Dictionary<String, Dictionary<String, Any>>
        let map: ConfidenceStruct = dataMap.mapValues { wrappedValue in
            let type = wrappedValue["type"] as! String
            return convertValue(type, wrappedValue["value"]!)
        }
        return ConfidenceValue.init(structure: map)
    case "list":
        let list = value as! [Dictionary<String, Any>]
        return ConfidenceValue.init(list: list.map { wrappedValue in
            let type = wrappedValue["type"] as! String
            return convertValue(type, wrappedValue["value"]!)
        })
    case "string":
        return ConfidenceValue.init(string: value as! String)
    case "unknown":
        // Dart could not map this type (a DateTime, a null, a custom object)
        // and has already sent `value.toString()`. Keep that string: coercing
        // it to a number would publish a wrong value with nothing to show the
        // caller their data was discarded.
        return ConfidenceValue.init(string: value as? String ?? String(describing: value))
    default:
        // An unrecognised type marker means the Dart and native sides have
        // drifted. Preserve the value as a string and make the mismatch
        // visible rather than silently publishing a number.
        NSLog("%@", "Confidence SDK: unsupported value type '\(type)', publishing it as a string")
        return ConfidenceValue.init(string: value as? String ?? String(describing: value))
    }
}
